### PR TITLE
feat(CLOUDDST-22189): add push-to-cdn task

### PR DESCRIPTION
**Pull request by pkhander**
JIRA: [CLOUDDST-22188](https://issues.redhat.com/browse/CLOUDDST-22188)

Dependent on the container created from - https://github.com/konflux-ci/release-service-utils/pull/153
Dependent on the task created from - https://github.com/konflux-ci/release-service-catalog/pull/455

### CODE REVIEW

**Comment by pkhander** [comment]
Hey @konflux-ci/release-service-maintainers, Looking to get some initial feedback on this task, related to [CLOUDDST-22184](https://issues.redhat.com/browse/).

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: Any reason to start with `4.1.0`? We typically start with `0.1.0`.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: This variable is created, but then the script doesn't use it and uses the `$(params.*)` reference anyway. So either this can be removed, or the env var should be used in the script.

But I think the RPA is probably not needed anyway - see my other comment.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: I'd suggest to pass the path to the data json file which will contain what you need - see other tasks for how it works, e.g. https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/create-pyxis-image/create-pyxis-image.yaml#L49

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: I'm not 100 %, but I don't think you'll be able to write to this location.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: So for cdn-qa, you will still use `EXODUS_GW_ENV=live`? Is that right?

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: What's this for? It doesn't seem to be used here for anything.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: You will use the release-service-utils image in this task, right? Same with the other step.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: I don't think you can access other tasks' results in a task. Instead, this should be passed as a parameter to the task and provided in the pipeline. Also, it shouldn't be an absolute path, but a relative path inside a workspace.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: Why create this new variable just to hold another variable's value?

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: It would be nice to also print what is being printed, e.g. `echo Destination dir: $destination_dir`

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: If the actual copying is gonna happen in another step, you will need to pass the path somehow - and I don't think using a task's result is an option. So I think you will have to save the path to a file in the workspace. But keep in mind that all pipelineruns share the same volume, so you'll need to use the subdirectory.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml:1]: I have a general question: From looking at this, it seems that only one file will be pushed. How will this work for multiple files? For example, if there are several components in the snapshot and each of them wants to push something. Shouldn't this support pushing multiple files at a time?

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/push-to-cdn.yaml]: Maybe this is leftover copied code, but I don't see any CR jsons being saved here

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/push-to-cdn.yaml]: This isn't the name of your task result

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/push-to-cdn.yaml]: As mentioned in the other task, just use the one fetched by collect-data instead of fetching it again

**Review by johnbieren — COMMENTED** [review]
Is this supposed to run inside the firewall or it does not need to?

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: cp/ps error from collect-data, updated, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Updated as such, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Stopped using RELEASE_PLAN_ADMISSION directly, updated to use dataPath instead.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: ack, I have updated to store that in the workspace 
`exodus_rsync_config_path="$(workspaces.data.path)/$(params.subdirectory)/exodus-rsync.conf"`

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: I have commented those lines out for now, I am not 100% sure how to access the secrets stored in the vault. 
exodus-{stage/prod}-cert.crt and exodus-{stage/prod}-privatekey.pem will be stored there I believe

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Updated as such, using binaries_dir parameter instead, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Updated as such, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml:1]: Correct, it should be supporting multiple files, I have updated the steps accordingly, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Updated as such, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: correct, cp/ps error, I have updated the description, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Yes, correct! for cdn-qa, we will be using the stage cert and key but env is still live

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Yes, For testing purposes I have added this image which is a clone of release-service-utils. I update that before the MR is ready to merge!

**Comment by pkhander** [comment]
> Is this supposed to run inside the firewall or it does not need to?

It does not need to!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: Nit: Other descriptions don't end with `.`, so this shouldn't either.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [reply tasks/push-to-cdn/push-to-cdn.yaml]: OK, sounds good, thanks.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Cool.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [reply tasks/push-to-cdn/push-to-cdn.yaml]: OK, thanks for confirming.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [inline tasks/push-to-cdn/push-to-cdn.yaml]: I'm not sure how this is supposed to work. Currently this will not not exist. The previous step does not write into this file. Instead, it only writes this file path to `$(results.destination_path.path)` - which seems confusing, because I would expect `$destination_path` to be written there instead.

Comment (mmalina) [reply tasks/push-to-cdn/push-to-cdn.yaml]: But also, the only result this task declares is `srcDestPath`.

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/push-to-cdn.yaml]: I am quite sure this will just be `.exodus.env` the data file doesn't have a spec and the data key isn't contained in it. See https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/apply-mapping/tests/test-apply-mapping.yaml#L27 for an example one

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/push-to-cdn.yaml]: I'm not sure we have any secrets in our public tasks (we do for the ones that run inside the firewall - see https://github.com/hacbs-release/app-interface-deployments/blob/main/internal-services/catalog/create-advisory-task.yaml#L54) @davidmogar any input on what should be done here?

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: update, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: update, ty!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: I thought I was writing content to that location on lines 115-117 and accessing it here similar to this (https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/collect-data/collect-data.yaml#L173)

Regarding `srcDestPath`, I updated that on line 116.

Trying to store key:value for src:dest which can be accessed here 
and another task (publish-to-cgw) later.

Comment (mmalina) [reply tasks/push-to-cdn/push-to-cdn.yaml]: OK, my bad, you're right you're writing to that file on line 117. Could both steps maybe use the same variable for that file? It would increase readability. One uses `OUTPUT_PATH`, but it's only the relative path inside the workspace, and the other uses `src_dest_json`, but it's the absolute path.

I thought I saw an instance of saving stuff to a result that doesn't exist, but maybe I was seeing things that didn't exist...

Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: Updated as such, ty!

I did change that on you in my last commit :)
https://github.com/konflux-ci/release-service-catalog/compare/0b90fc923c6754cf3456e75367fec25f04bcdf3b..294ae36286ad403cd874ad1fc8016b19fb1ccbad

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (davidmogar) [reply tasks/push-to-cdn/push-to-cdn.yaml]: There's no Vault integration. That's only available in internal pipelines as Johnny mentioned. The secret would have to be defined somehow in the workspace and then you could access it by defining an environment variable. Something like the following:
```
valueFrom:
  secretKeyRef:
    name: secret
    key: key
```

Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: ty for the pointer, Searching for such instance in the repo, I see [create-github-release.yaml](https://github.com/konflux-ci/release-service-catalog/blob/bbecccc99e319143eb80de49be856b11e7198665/tasks/create-github-release/create-github-release.yaml#L49) and [create-pyxis-image.yaml](https://github.com/konflux-ci/release-service-catalog/blob/bbecccc99e319143eb80de49be856b11e7198665/tasks/create-pyxis-image/create-pyxis-image.yaml#L66) using similar env variables. 

I see they both are dependent on additional tasks, [collect-gh-params](https://github.com/konflux-ci/release-service-catalog/blob/bbecccc99e319143eb80de49be856b11e7198665/pipelines/release-to-github/release-to-github.yaml#L290) and [collect-pyxis-parms](https://github.com/konflux-ci/release-service-catalog/blob/bbecccc99e319143eb80de49be856b11e7198665/tasks/collect-pyxis-params/collect-pyxis-params.yaml) 
Is a separate task preferred here as well, something like `collect-exodus-params` ?

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (mmalina) [reply tasks/push-to-cdn/push-to-cdn.yaml]: ok, cool!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [reply tasks/push-to-cdn/push-to-cdn.yaml]: The new image is available here quay.io/konflux-ci/release-service-utils:38c3bfd7479c86b832cba5b61f9bbde40c469393

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [reply tasks/push-to-cdn/push-to-cdn.yaml]: This is because we have different secrets if we are running in stage or prod, and you can't determine what secret you want and then attach it to the pod in the same task. If this also will run against either the stage or prod instance, yes, we will need something like `collect-exodus-params` unfortunately

Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml]: I have created this MR for a new task - `prepare-exodus-params` (since there was bit more processing using _prepare_ instead of _collect_)  - https://github.com/konflux-ci/release-service-catalog/pull/455 

and updated this PR to use those params instead

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/push-to-cdn.yaml:1]: You have `set -x` in the task but you are writing keys and certs. Can you adjust? Ideally so that set -x is still called when things are being done, but not when the keys and certs are being written

File: tasks/push-to-cdn/tests/pre-apply-task-hook.sh
Comment (johnbieren) [inline tasks/push-to-cdn/tests/pre-apply-task-hook.sh:8]: Can you adjust the secret name to be something more related to the task name? If we have multiple tasks with exodus secrets, we can run into issues here when running CI

File: tasks/push-to-cdn/tests/test-push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/tests/test-push-to-cdn.yaml]: Can you adjust this to a konflux-ci/release-service-utils image? Any other task should have one to copy from (including your check-result below)

File: tasks/push-to-cdn/tests/test-push-to-cdn.yaml
Comment (johnbieren) [inline tasks/push-to-cdn/tests/test-push-to-cdn.yaml]: needs update per new secret name

Comment (pkhander) [reply tasks/push-to-cdn/tests/test-push-to-cdn.yaml]: aha, updated, thanks!

File: tasks/push-to-cdn/push-to-cdn.yaml
Comment (pkhander) [reply tasks/push-to-cdn/push-to-cdn.yaml:1]: Updated as such, ty!

**Review by johnbieren — APPROVED** [review]

**Review by johnbieren — APPROVED** [review]
/ok-to-test

**Comment by johnbieren** [comment]
@pkhander do you want me to merge this or do you have more changes to make to it?

**Comment by pkhander** [comment]
> @pkhander do you want me to merge this or do you have more changes to make to it?

@johnbieren This is ready to be merged, no changes left to make as of now, ty!

**Pull request merged**
